SetFilter: support set expressions

mattigruener opened the pull request:
Hey John,

this is a first, naive stab at adding support for set expressions on that filter node. I'm creating the PR already because I have a question about it. Will do that as a comment.

Thanks :)

Matti.

mattigruener commented on src/GafferScene/SetAlgo.cpp:
```diff
@@ -308,6 +318,11 @@ struct AstHasher
 
+	void operator()( const Nil &nil )
+	{
+		m_hash.append( false );
```
One of the tests in `SetFilterTest` runs the node with an empty string as the set name/expression. I had to explicitly hash a dummy value here, so that Gaffer doesn't throw an exception at me because `SetFilter::hash()` never actually hashes anything in that case. Do you have a better idea?

johnhaddon replied:
> One of the tests in SetFilterTest runs the node with an empty string as the set name/expression. I had to explicitly hash a dummy value here, so that Gaffer doesn't throw an exception at me because SetFilter::hash() never actually hashes anything in that case. Do you have a better idea?

I don't think the SetAlgo change is the answer here. The clue is in the [documentation for ComputeNode::hash()](https://github.com/GafferHQ/gaffer/blob/master/include/Gaffer/ComputeNode.h#L68) - you need to make sure you call your base class hash before appending your own, in which case it's fine to append nothing. The base class hash will end up at the top of the hierarchy in [`ComputeNode::hash()`](https://github.com/GafferHQ/gaffer/blob/master/src/Gaffer/ComputeNode.cpp#L56), where the TypeId for the node and the name of the plug are used to start the hash with something unique to the node/plug combo.

> I guess we should create an empty PathMatcher for these cases after all? ;) You've been saying that all along, I think.

Oh, I thought we were doing! Yes, given that the SetFilter has been doing that all along without any complaint from users, I think we can say that it's the right thing to do. Note that this means you can entirely remove the `setNames()` check from SetAlgo, since it's the responsibility of the `SceneNode::computeSet()` to return an empty set where the name is invalid anyway.

mattigruener commented:
Hey John,

I think the two things we talked about should be fixed now.

One thing I noticed and that I found confusing - maybe that's just me though - was that I got crashes until I removed the expressionResultPlug from the UI. The reason was that computation for that plug was triggered and the set expression was evaluated and accessed the scene which was invalid, because it wasn't set up by `FilteredSceneProcessor` in that case. I've removed it from the UI now of course and I've also added a check in SetAlgo to avoid hard crashes in the future. It did confuse me, though - because triggering the compute it's not really needed at that point I guess? I guess I expected filters to only be triggered when a scene node pulled on its filter plug... Anyway, just a remark as a "user" of the API. Hunting this down definitely was a good exercise to learn more about how the filtering works in gaffer :)

How does this look to you now? I didn't rename the plug that holds the set name/expression. It seems better like this in terms of backward compatibility and such? You didn't seem to have a clear preference?

I haven't done anything special for the case of a single set name being used as the expression, yet. Do you want me to address your concern about performance in this PR or once we notice an actual decrease in speed?

Thanks :)

Matti.

johnhaddon commented:
Hi Matti,
I'm logging off now so won't have a chance to look through this properly till tomorrow. Just wanted to give you a heads up that I think we should do some performance testing before merging - it's easier to do that now than to track down problems once they've appeared in production. I think some sort of artificial test that is heavy on the filter would be ideal - maybe calling `SceneAlgo::matchingPaths()` for a large scene?
Cheers...
John

mattigruener commented:
Ok great, I will look into that after lunch. Thanks :)

johnhaddon commented on src/GafferScene/SetFilter.cpp:
```diff
@@ -56,6 +57,7 @@ SetFilter::SetFilter( const std::string &name )
 	storeIndexOfNextChild( g_firstPlugIndex );
 
 	addChild( new StringPlug( "set" ) );
+	addChild( new PathMatcherDataPlug( "expressionResult", Gaffer::Plug::Out, new PathMatcherData ) );
```
We should prefix the name with __ to show that it's private.

johnhaddon commented on src/GafferScene/SetFilter.cpp:
```diff
@@ -121,6 +150,17 @@ void SetFilter::hashMatch( const ScenePlug *scene, const Gaffer::Context *contex
 	h.append( scene->setHash( setPlug()->getValue() ) );
 }
 
+void SetFilter::compute( Gaffer::ValuePlug *output, const Gaffer::Context *context ) const
```
Put this just below `SetFilter::hash()`, so it's easy to verify the two functions match. Making sure that hash/compute have matching behaviour is one of the most irksome things about coding for Gaffer, and keeping them together does help a bit.

johnhaddon commented on src/GafferScene/SetFilter.cpp:
```diff
@@ -129,7 +169,12 @@ unsigned SetFilter::computeMatch( const ScenePlug *scene, const Gaffer::Context
 	const ScenePlug::ScenePath &path = context->get<ScenePlug::ScenePath>( ScenePlug::scenePathContextName );
-	ConstPathMatcherDataPtr set = scene->set( setPlug()->getValue() );
+
+	Gaffer::ContextPtr pathMatcherContext = new Gaffer::Context( *context, Gaffer::Context::Borrowed );
```
Let's use the fancy new ScenePlug::GlobalScope here.

We've also got a mismatch between the `computeMatch()` and `hashMatch()` methods now - we need to hash in `expressionResultPlug()` in `hashMatch()`, rather than calling `setHash()`. We'll need to use a GlobalScope for that too.

mattigruener replied:
> Let's use the fancy new ScenePlug::GlobalScope here.

We can't, can we? That one not only throws away `scene:path` but also the scene that is needed for creating the `ExpressionAst`?

https://github.com/GafferHQ/gaffer/blob/04fa7b2b729eb18e4012603bfd46c05b19bbb92e/src/GafferScene/ScenePlug.cpp#L287

johnhaddon replied:
Good point. Let's use the fancy new `Context::EditableScope` then...

johnhaddon commented on python/GafferSceneUI/SetFilterUI.py:
```diff
@@ -57,15 +57,22 @@
 
 		],
 
+		"expressionResult" : [
```
Shouldn't need any of this once the name has an __ prefix.

johnhaddon commented on python/GafferSceneUI/SetFilterUI.py:
```diff
@@ -57,15 +57,22 @@
 			"""
-			The name of a set that defines the locations to
-			be matched.
+			A set expression that computes a set that defines
```
Could you include a simple example of an expression here, and mention that a kind soul has also provided extensive documentation of the syntax in the reference section?

johnhaddon commented:
Hi Matti,

Not sure exactly what's going on, but I think my review and some changes you've pushed have coincided at the same time. All my comments were "pending" and I was about to write the final review comment but not GitHub shows all the comments as published - hopefully you see them? A couple are "show outdated" ones now, but I believe at least one of those is still relevant.

> One thing I noticed and that I found confusing - maybe that's just me though - was that I got crashes until I removed the expressionResultPlug from the UI....did confuse me, though - because triggering the compute it's not really needed at that point I guess?

I suspect it was something like the label for the plug, calling `getValue()` to see if the value is at the default or not, and show that little green asterisk.

> I didn't rename the plug that holds the set name/expression. It seems better like this in terms of backward compatibility and such? You didn't seem to have a clear preference?

We could deal with backwards compatibility with a pretty simple config file. I quite like the idea of renaming the plug to "expression", because that gives folks a pretty big hint that they can put stuff in there that they couldn't before. Maybe see what @andrewkaufman, @danieldresser and @donboie think and make a call?

> I haven't done anything special for the case of a single set name being used as the expression, yet. Do you want me to address your concern about performance in this PR or once we notice an actual decrease in speed?

I'm hoping we won't need a special case for single sets, but I think there's a bit more testing to be done (especially once you've fixed the bug in hashMatch). I'll reply offline about those specific tests...

Cheers...
John

mattigruener commented:
Sorry, I guess I started addressing your feedback a little overly excited and should've given you more time to finish it. I will go through all of your comments and make sure they're addressed.

mattigruener commented:
Things left to do (just so you know I have that on my list):
- fix that test so that it doesn't check for `RuntimeErrors` anymore
- rename the plug (Andrew also is in favor of doing that)
- create compatibility config
- fix the UI
  - selecting set should probably just insert it at the cursor instead of replacing the plug value
  - add sub menu called operator that contains these: Intersection (&), Union (|) and Difference (-)
  - selecting one of those would insert "A & B", "A | B", "A - B"

The UI stuff is the result of having a quick chat with Andrew and Don about it. Does that make sense to you? :) Would we wrap the "A OP B" in parentheses? Andrew brought up that operator menu because it's a nice way to learn the available operators (similar to how the OSLCode node offers all OSL functions in a menu).

johnhaddon commented:
Hey Matti,

Thanks for the update. Sorry I haven't got to looking at this today - slogging away at this Box UI stuff instead, since that seems to be the main blocker in getting a release out.

It's interesting that you called the `-` operator "Difference". I've been thinking that ANDNOT (as it's called in the code and the docs) was a strange terminology but hadn't got round to mentioning it. Maybe you could update the reference docs to use the Intersection/Union/Difference terms too, instead of AND/OR/ANDNOT?

I like the idea of having the operators in the menu to help people discover them. Not sure about the "parenthesis or not?" question - I think my preference might be to simply insert the operator, without the A or the B even, since I'd just have to delete the A and the B anyway. Maybe the menu item could say "Intersection (A & B)" to demonstrate that operands are needed, but then just insert the '&' into the text?

Cheers...
John

mattigruener commented:
`I like the idea of having the operators in the menu to help people discover them. Not sure about the "parenthesis or not?" question - I think my preference might be to simply insert the operator, without the A or the B even, since I'd just have to delete the A and the B anyway. Maybe the menu item could say "Intersection (A & B)" to demonstrate that operands are needed, but then just insert the '&' into the text?`

Sounds good to me too. We thought adding the operands would then lead to people selecting them, right-clicking, choosing a set and having the operand replace by a proper name, but maybe it's enough to just insert the operators :)

I'll update the docs regarding the "Difference" terminology. I was thinking about the right term to use and looked at [this one](http://www.beva.org/maen503/week2/sets.htm) for reference.

Looking at that list, we're currently "missing" a complement operator. I guess for light linking we could easily have one (and use __lights as "everything"), but for every other usage it would be difficult and we should probably just ignore it...

mattigruener commented:
Hey John :)

Thanks for the feedback - I've addressed it in the latest commits. The UI should be updated now, too. About that one: I introduced a "ui:scene:acceptsSetExpression" metadata entry to complement what we had already. That should be ok? The code in `SetUI` got a little uglier than it was before, because the new functionality is not entirely in line with what we had. Let me know if you want me to have another stab at it. I just tried to squeeze the new stuff into what we had already... 

The other changes should be fairly trivial. Do we want tests for compatibility configs to make sure the old plug name is still usable in .gfr scripts?

Not sure what the convention is, but I've also changed american spelling to english spelling somewhere in the docs - I believe you use BE everywhere.

Seems this is getting pretty close to be mergeable :)

Thanks again,

Matti.

mattigruener commented:
> Ensure that the tests pass without the config. This is the step I think we're missing at the moment.

Yes, you're right. I just tried running the tests and get all kinds of failures because my env is in a super messed up state. I'll try to clean that up and get everything to pass without the config - thanks for pointing that out.

I have added commits to address the other feedback.

mattigruener commented:
The last commit simply renames the plug in some of the tests. I didn't see any other failures that looked related.

johnhaddon commented:
Thanks Matti, I've merged this after squashing via #2052...
